### DISCUSSION TITLE
12% speed-up to vector incremental builds

### DISCUSSION THREAD

**Discussion by kapilsinha** [vectordotdev/vector#19930] [General]
# 12% speed-up to vector.dev incremental builds
## Summary
I have forked & modified the Rust compiler, achieving 12% faster incremental builds for vector.dev. The modded compiler can cache proc macro expansion; as you know, heavy proc macro dependencies impact compilation time and cause lags in rust-analyzer (in VSCode). vector.dev depends on proc-macro heavy crates like async-graphql, and thus benefits from this optimization. I have profiled vector's incremental builds, and have found that with zero configuration I achieve 12% faster builds (via cargo check):

![vector_build_speedup](https://github.com/vectordotdev/vector/assets/13107230/5b12dd2e-00c6-4c19-96c3-902a351defa1)

## More details
I learned that the Rust compiler re-expands every macro in a crate, even on one-line edits, which is a disaster for the edit-compile-run cycle. Taking a similar approach to Rust's incremental compilation, I modified the Rust compiler to cache these expansions (in particular, the ones without side-effects; the `counter!` macros in vector.dev are the only ones that have side-effects) so that we don't need to re-compute them when unrelated code changes. This proc macro cache is the first tool in a toolkit that I am building -- all with the goal of improving Rust developer productivity!

## Chat?
I'd love to chat and/or have you try it out! I've recently packaged it into a remote build tool, which allows you to try it out in <3 minutes.

**Comment by jszwedko** [comment]
Thanks for sharing @kapilsinha ! That is interesting and quite the improvement. I don't think we'd want to run a remote build tool though, do you have any plans to upstream these changes to Rust?

**Reply by kapilsinha** [reply]
Hey @jszwedko thanks for taking a look! I’m still relatively early in figuring out how best to deliver my tool to end users:

The remote build toolkit has the advantage of being bundled with other tools I’m working on (eg the ability to change git branches and automatically maintain different build directories ie avoid rebuilds) and allowing a dev team to share caches on a beefy machine. I do think this lends itself to the best possible dev experience but I am open to creating a self-hostable version if that works better. 

I am also experimenting with just shipping my compiler directly (easily installable as a rustup toolchain). Frankly, I’m a bit hesitant to try to merge this into Rust itself (for sake of delivery speed and monetizability) — but I’d love to have a brief chat with you / your team to understand your needs!

**Reply by kapilsinha** [reply]
[Edit] Thanks for the chat! Very helpful getting your perspective on faster builds for Rust and especially open-source
